// scavio/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  BadRequestError: () => BadRequestError,
  InsufficientCreditsError: () => InsufficientCreditsError,
  InvalidAPIKeyError: () => InvalidAPIKeyError,
  MissingAPIKeyError: () => MissingAPIKeyError,
  RateLimitError: () => RateLimitError,
  Scavio: () => Scavio,
  ScavioAPIError: () => ScavioAPIError,
  ScavioError: () => ScavioError
});
module.exports = __toCommonJS(index_exports);

// src/errors.ts
var ScavioError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "ScavioError";
  }
};
var MissingAPIKeyError = class extends ScavioError {
  constructor() {
    super(
      "No API key provided. Pass apiKey or set the SCAVIO_API_KEY environment variable. Get your free key at https://dashboard.scavio.dev"
    );
    this.name = "MissingAPIKeyError";
  }
};
var InvalidAPIKeyError = class extends ScavioError {
  constructor(message = "Invalid API key") {
    super(message);
    this.name = "InvalidAPIKeyError";
  }
};
var InsufficientCreditsError = class extends ScavioError {
  constructor(message = "Insufficient credits") {
    super(message);
    this.name = "InsufficientCreditsError";
  }
};
var BadRequestError = class extends ScavioError {
  constructor(message = "Bad request") {
    super(message);
    this.name = "BadRequestError";
  }
};
var RateLimitError = class extends ScavioError {
  constructor(message = "Rate limit exceeded") {
    super(message);
    this.name = "RateLimitError";
  }
};
var ScavioAPIError = class extends ScavioError {
  statusCode;
  constructor(statusCode, message) {
    super(`API error ${statusCode}: ${message}`);
    this.name = "ScavioAPIError";
    this.statusCode = statusCode;
  }
};

// src/http.ts
var BASE_URL = "https://api.scavio.dev";
var DEFAULT_TIMEOUT = 3e4;
function buildHeaders(apiKey) {
  return {
    Authorization: `Bearer ${apiKey}`,
    "Content-Type": "application/json",
    "X-Client-Source": "scavio-js"
  };
}
function stripUndefined(obj) {
  const result = {};
  for (const [key, value] of Object.entries(obj)) {
    if (value !== void 0) {
      result[key] = value;
    }
  }
  return result;
}
function handleError(statusCode, body) {
  let error = body.error ?? "Unknown error";
  if (typeof error === "object" && error !== null && "message" in error) {
    error = error.message;
  }
  const msg = String(error);
  if (statusCode === 400) throw new BadRequestError(msg);
  if (statusCode === 401) throw new InvalidAPIKeyError(msg);
  if (statusCode === 402) throw new InsufficientCreditsError(msg);
  if (statusCode === 429) throw new RateLimitError(msg);
  throw new ScavioAPIError(statusCode, msg);
}
async function request(options) {
  await options.rateLimiter.wait();
  const url = `${options.baseUrl}${options.path}`;
  const headers = buildHeaders(options.apiKey);
  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), options.timeout);
  try {
    const fetchOptions = {
      method: options.method,
      headers,
      signal: controller.signal
    };
    if (options.method === "POST" && options.body) {
      fetchOptions.body = JSON.stringify(stripUndefined(options.body));
    }
    const response = await fetch(url, fetchOptions);
    if (!response.ok) {
      let body = {};
      try {
        body = await response.json();
      } catch {
      }
      handleError(response.status, body);
    }
    return await response.json();
  } finally {
    clearTimeout(timeoutId);
  }
}

// src/rate-limiter.ts
var RateLimiter = class {
  maxPerSecond;
  timestamps = [];
  pending = Promise.resolve();
  constructor(maxPerSecond) {
    this.maxPerSecond = maxPerSecond;
  }
  async wait() {
    const ticket = this.pending.then(() => this.acquire());
    this.pending = ticket;
    return ticket;
  }
  async acquire() {
    this.cleanup();
    if (this.timestamps.length >= this.maxPerSecond) {
      const sleepMs = 1e3 - (Date.now() - this.timestamps[0]);
      if (sleepMs > 0) {
        await new Promise((resolve) => setTimeout(resolve, sleepMs));
      }
      this.cleanup();
    }
    this.timestamps.push(Date.now());
  }
  cleanup() {
    const now = Date.now();
    while (this.timestamps.length > 0 && now - this.timestamps[0] >= 1e3) {
      this.timestamps.shift();
    }
  }
};

// src/namespaces/amazon.ts
var AmazonNamespace = class {
  constructor(client) {
    this.client = client;
  }
  client;
  async search(options) {
    return this.client._post("/api/v1/amazon/search", options);
  }
  async product(options) {
    const { asin, ...rest } = options;
    return this.client._post("/api/v1/amazon/product", {
      query: asin,
      ...rest
    });
  }
};

// src/namespaces/google.ts
var GoogleNamespace = class {
  constructor(client) {
    this.client = client;
  }
  client;
  async search(options) {
    return this.client._post("/api/v1/google", options);
  }
};

// src/namespaces/reddit.ts
var RedditNamespace = class {
  constructor(client) {
    this.client = client;
  }
  client;
  async search(options) {
    return this.client._post("/api/v1/reddit/search", options);
  }
  async post(options) {
    return this.client._post("/api/v1/reddit/post", options);
  }
};

// src/namespaces/tiktok.ts
var TikTokNamespace = class {
  constructor(client) {
    this.client = client;
  }
  client;
  async profile(options) {
    return this.client._post("/api/v1/tiktok/profile", options);
  }
  async userPosts(options) {
    return this.client._post("/api/v1/tiktok/user/posts", options);
  }
  async video(options) {
    return this.client._post("/api/v1/tiktok/video", options);
  }
  async videoComments(options) {
    return this.client._post("/api/v1/tiktok/video/comments", options);
  }
  async commentReplies(options) {
    return this.client._post("/api/v1/tiktok/video/comments/replies", options);
  }
  async searchVideos(options) {
    return this.client._post("/api/v1/tiktok/search/videos", options);
  }
  async searchUsers(options) {
    return this.client._post("/api/v1/tiktok/search/users", options);
  }
  async hashtag(options) {
    return this.client._post("/api/v1/tiktok/hashtag", options);
  }
  async hashtagVideos(options) {
    return this.client._post("/api/v1/tiktok/hashtag/videos", options);
  }
  async userFollowers(options) {
    return this.client._post("/api/v1/tiktok/user/followers", options);
  }
  async userFollowings(options) {
    return this.client._post("/api/v1/tiktok/user/followings", options);
  }
};

// src/namespaces/instagram.ts
var InstagramNamespace = class {
  constructor(client) {
    this.client = client;
  }
  client;
  async profile(options) {
    return this.client._post("/api/v1/instagram/profile", options);
  }
  async userPosts(options) {
    return this.client._post("/api/v1/instagram/user/posts", options);
  }
  async userReels(options) {
    return this.client._post("/api/v1/instagram/user/reels", options);
  }
  async userTagged(options) {
    return this.client._post("/api/v1/instagram/user/tagged", options);
  }
  async userStories(options) {
    return this.client._post("/api/v1/instagram/user/stories", options);
  }
  async post(options) {
    return this.client._post("/api/v1/instagram/post", options);
  }
  async postComments(options) {
    return this.client._post("/api/v1/instagram/post/comments", options);
  }
  async commentReplies(options) {
    return this.client._post("/api/v1/instagram/post/comments/replies", options);
  }
  async searchUsers(options) {
    return this.client._post("/api/v1/instagram/search/users", options);
  }
  async searchHashtags(options) {
    return this.client._post("/api/v1/instagram/search/hashtags", options);
  }
  async userFollowers(options) {
    return this.client._post("/api/v1/instagram/user/followers", options);
  }
  async userFollowings(options) {
    return this.client._post("/api/v1/instagram/user/followings", options);
  }
};

// src/namespaces/walmart.ts
var WalmartNamespace = class {
  constructor(client) {
    this.client = client;
  }
  client;
  async search(options) {
    return this.client._post("/api/v1/walmart/search", options);
  }
  async product(options) {
    return this.client._post("/api/v1/walmart/product", options);
  }
};

// src/namespaces/youtube.ts
var YouTubeNamespace = class {
  constructor(client) {
    this.client = client;
  }
  client;
  async search(options) {
    const { query, ...rest } = options;
    return this.client._post("/api/v1/youtube/search", {
      search: query,
      ...rest
    });
  }
  async metadata(options) {
    return this.client._post("/api/v1/youtube/metadata", options);
  }
};

// src/client.ts
var Scavio = class {
  google;
  amazon;
  walmart;
  youtube;
  reddit;
  tiktok;
  instagram;
  apiKey;
  baseUrl;
  timeout;
  rateLimiter;
  constructor(config) {
    this.apiKey = config?.apiKey ?? process.env.SCAVIO_API_KEY ?? "";
    if (!this.apiKey) {
      throw new MissingAPIKeyError();
    }
    this.baseUrl = (config?.baseUrl ?? BASE_URL).replace(/\/+$/, "");
    this.timeout = config?.timeout ?? DEFAULT_TIMEOUT;
    const rps = config?.maxRequestsPerSecond ?? 1;
    if (rps < 1 || rps > 10) {
      throw new ScavioError("maxRequestsPerSecond must be between 1 and 10");
    }
    this.rateLimiter = new RateLimiter(rps);
    this.google = new GoogleNamespace(this);
    this.amazon = new AmazonNamespace(this);
    this.walmart = new WalmartNamespace(this);
    this.youtube = new YouTubeNamespace(this);
    this.reddit = new RedditNamespace(this);
    this.tiktok = new TikTokNamespace(this);
    this.instagram = new InstagramNamespace(this);
  }
  /** @internal */
  async _post(path, body) {
    return request({
      method: "POST",
      path,
      apiKey: this.apiKey,
      baseUrl: this.baseUrl,
      timeout: this.timeout,
      rateLimiter: this.rateLimiter,
      body
    });
  }
  /** @internal */
  async _get(path) {
    return request({
      method: "GET",
      path,
      apiKey: this.apiKey,
      baseUrl: this.baseUrl,
      timeout: this.timeout,
      rateLimiter: this.rateLimiter
    });
  }
  async search(options) {
    return this.google.search(options);
  }
  async getUsage() {
    return this._get("/api/v1/usage");
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BadRequestError,
  InsufficientCreditsError,
  InvalidAPIKeyError,
  MissingAPIKeyError,
  RateLimitError,
  Scavio,
  ScavioAPIError,
  ScavioError
});
